lib/server.js

```diff
@@ -0,0 +1,205 @@
+import http from 'node:http'
+import net from 'node:net'
+import {EventEmitter} from 'node:events'
+import crypto from 'node:crypto'
+import url from 'node:url'
+
+import {Headers} from 'undici'
+
+import {SafeHttpClient, HttpError} from './safe-http-client.js'
+import {
+  DEFAULT_HEADERS,
+  REQUEST_HEADERS,
+  RESPONSE_HEADERS
+} from './constants.js'
+
+/**
+ * @typedef {Object} Config
+ * @property {string} HMACKey
+ * @property {string} serverName Server name used in Headers and Via checks
+ * @property {number|undefined} [maxSize] The max size in bytes per resource to download. Sends 500 if exceeded.
+ */
+
+/**
+ * @param {string} hex
+ */
+function hexDecode(hex) {
+  const bytes = []
+  for (let i = 0; i < hex.length; i += 2) {
+    bytes.push(parseInt(hex.substr(i, 2), 16))
+  }
+  return Buffer.from(bytes)
+}
+
+/**
+ * Filter incoming headers based on allowed headers.
+ * We always want to send Pascal-Case headers so we use the key from constants
+ * with the value from the incoming header.
+ * @param {Map<string, boolean>} allowedHeaders
+ * @returns {(incomingHeaders: Headers) => Record<string, string>}
+ * */
+function filterHeaders(allowedHeaders) {
+  return function (incomingHeaders) {
+    /** @type {Record<string, string>} */
+    const filteredHeaders = {}
+
+    for (const [key, bool] of allowedHeaders.entries()) {
+      if (incomingHeaders.has(key.toLowerCase()) && bool) {
+        // @ts-ignore
+        filteredHeaders[key] = incomingHeaders.get(key.toLowerCase())
+      }
+    }
+
+    return filteredHeaders
+  }
+}
+
+/*
+ * Node.js HTTP server to proxy insecure content via HTTPS.
+ */
+export class Server extends EventEmitter {
+  /**
+   * @param {Config} config
+   */
+  constructor(config) {
+    super()
+    if (!config.HMACKey) {
+      throw new Error('HMACKey is required')
+    }
+    this.config = config
+    this.config.serverName ??= 'camo'
+  }
+
+  /**
+   * Start the server. Identical to `net.Server.listen()`.
+   * @param {Parameters<InstanceType<typeof net.Server>['listen']>} args
+   * @public
+   */
```

Comment:
```suggestion
  /**
   * Start the server.
   * 
   * @param {Parameters<InstanceType<typeof net.Server>['listen']>} args
   *   Arguments passedf to `net.Server.listen`.
   * @returns {net.Server}
   *   Server.
   */
```

a) docs, b) `@public` is the default

Q: Should it be possible to stop the server?

Reply:
Return type is already inferred and it's [generally the consensus to not add it unnecessarily](https://www.youtube.com/watch?v=nwSe95uFN8E).

You can do whatever you want with the server in the same way as you would when using `net.Server`

```js
const camomile = new Server({HMACKey})
const server = camomile.listen()

await server.close()
```

Reply:
The linked video sparked a lot of discussion on Twitter. There isn’t really a general consensus on this topic for application code. For library code it’s different.

[A quote from the video](https://www.youtube.com/watch?v=nwSe95uFN8E&t=152s):

> If you’re writing library code, you should always declare return types.

This gives you more control over the types that TypeScript will generate.

b) Let’s remove `@public`. Methods are public by default.